bff/src/main/java/ksch/bff/LoginController.java

```diff
@@ -0,0 +1,59 @@
+package ksch.bff;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import kong.unirest.Unirest;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
+
+@RestController
+public class LoginController {
+
+    private static final String TOKEN_URL = "https://noauth-ga2speboxa-ew.a.run.app/oauth/token";
+
+    @GetMapping("/bff/callback")
+    ResponseEntity<?> handleAuthorizationCallback(@RequestParam String code, HttpServletRequest request) {
+
+        var tokenResponse = exchangeAuthorizationGrant(code);
+
+        var session = request.getSession();
+        session.setAttribute("accessToken", tokenResponse.getAccessToken());
+        session.setAttribute("refreshToken", tokenResponse.getRefreshToken());
+
+        return ResponseEntity.ok().build();
```

Comment:
Redirect to originally requested page

Reply:
```mermaid
sequenceDiagram
    autonumber
    participant SPA
    participant BFF
    participant AS as Authorization Server
    
    BFF->SPA: Interception
    BFF->>AS: Redirect
    AS->>BFF: Redirect
    BFF->>AS: Request access token
    AS-->>BFF: Access token
    BFF->>SPA: Redirect

```

Reply:
(1) The BFF intercepts the GET requests for the HTML code of the SPA. Then it checks whether the user session has an access token.

(2) If the user session does not have an access token, the user's browser is redirected to the authorization server. To be able to redirect back to the SPA at the end of the OAuth process, the intercepted URL is stored in the user session.

(3) After the users have granted access, the authorization server redirects the user back to the BFF, with the authorization code as query parameter.

(4) With the help of the authorization code and its client ID and client secret, the BFF requests an access token from the authorization server.

(5) The response contains the access and refresh token which get stored in the user session.

(6) Now the users can be redirected back to the SPA which they originally requested. When the SPA makes an HTTP request to the BFF, then it will always include the session token. The session token is the key to the access token. The BFF adds the access token to the request and then forwards the request to the backend which then checks the access token.